// @rmp135/sql-ts/dist/Adapters/sqlite.js
import * as SharedAdapterTasks from './SharedAdapterTasks.js';
export default {
    async getAllEnums(db, config) {
        return await SharedAdapterTasks.getTableEnums(db, config);
    },
    async getAllTables(db, schemas) {
        const sql = `
      SELECT tbl_name from sqlite_master
      WHERE tbl_name <> 'sqlite_sequence'
      AND type IN ('table', 'view')
    `;
        return (await db.raw(sql))
            .map((t) => ({ name: t.tbl_name, schema: 'main', comment: '' }));
    },
    async getAllColumns(db, config, table, schema) {
        return (await db.raw(`pragma table_info(${table})`))
            .map((c) => ({
            name: c.name,
            nullable: c.notnull === 0,
            type: (c.type.includes('(') ? c.type.split('(')[0] : c.type).toLowerCase(),
            optional: c.dflt_value !== null || c.notnull === 0 || c.pk !== 0,
            columnType: 'Standard',
            isPrimaryKey: c.pk !== 0,
            comment: '',
            defaultValue: c.dflt_value?.toString() ?? null,
        }));
    }
};
